src/java.base/share/classes/jdk/crac/RCULock.java

```diff
@@ -0,0 +1,324 @@
+    private SwitchPoint lockSwitchPoint = new SwitchPoint();
+    private SwitchPoint unlockSwitchPoint = new SwitchPoint();
+    private volatile MethodHandle lockImpl;
+    private volatile MethodHandle unlockImpl;
+
+    static {
+        try {
+            Field readerThreadsList = RCULock.class.getDeclaredField("readerThreadsList");
+            Field readCriticalMethods = RCULock.class.getDeclaredField("readCriticalMethods");
+            initFieldOffsets(readerThreadsList, readCriticalMethods);
+        } catch (NoSuchFieldException e) {
+            throw new ExceptionInInitializerError(e);
+        }
+    }
+
+    private static native void initFieldOffsets(Field readerThreadsList, Field readCriticalMethods);
+
+    /**
+     * Inspect the classes for all methods (including private ones) marked with
+     * the {@link RCULock.Critical} annotation and create a lock that can protect
+     * these methods with read lock.
+     *
+     * @param classes List of classes to be inspected.
+     * @return New lock instance.
+     */
+    public static RCULock forClasses(Class<?>... classes) {
+        return new RCULock(findAnnotated(classes));
+    }
+
+    private static List<Method> findAnnotated(Class<?>[] classes) {
+        List<Method> methods = new ArrayList<>();
+        for (Class<?> cls : classes) {
+            while (cls != null && cls != Object.class) {
+                for (Method m : cls.getDeclaredMethods()) {
+                    if (m.isAnnotationPresent(Critical.class)) {
+                        methods.add(m);
+                    }
+                }
+                cls = cls.getSuperclass();
+            }
+        }
+        return methods;
+    }
+
+    /**
+     * Create a lock that can protect selected methods with read-locking.
+     *
+     * @param readCriticalMethods List of methods with the read-critical section.
+     */
+    public RCULock(Iterable<Method> readCriticalMethods) {
+        this(StreamSupport.stream(readCriticalMethods.spliterator(), false)
+                .map(RCULock::signature).toArray(String[]::new));
+    }
+
+    private static String signature(Method m) {
+        return m.getDeclaringClass().getName() + "." + m.getName() + "(" +
+                Arrays.stream(m.getParameterTypes())
+                        .map(Class::descriptorString).collect(Collectors.joining())
+                + ")" + m.getReturnType().descriptorString();
+    }
+
+    /**
+     * Create a lock that can protect selected methods with read-locking.
+     * <p>
+     * The signatures follow the form
+     * <pre>my.package.MyClass.foobar(another/package/SomeClass;Z)V</pre>
+     *
+     * @apiNote This constructor does not use var-args argument to prevent
+     * accidentally creating a RCULock without declaring any critical methods.
+     *
+     * @param readCriticalMethods List of signatures for methods invoked in the read-critical section.
+     */
+    public RCULock(String[] readCriticalMethods) {
```

Comment:
It's a big question, why (from the interface point of view) the lock should know methods in which it can be used.

The second tier question, why the method names are String[] and not Method[] at least.

Reply:
There's no point in asking 'why'. Interface is there to specify a contract. You declare the methods, and then the implementation fulfills the contract. And you already know why from the point of implementation.

There is another constructor that accepts `Method[]`, being able to specify the signature is just an option if it's more convenient.